File: movile_front/lib/services/rental_vehicles_service.dart
import 'package:autogo_mobile/models/models.dart';
import 'api_service.dart';

class RentalVehiclesService {
  final ApiService _apiService = ApiService();

  Future<List<RentalVehicle>> getRentalVehicles({String? token}) async {
    try {
      final response = await _apiService.get('/rental-vehicles', token: token);
      if (response is List) {
        return response
            .map((json) => RentalVehicle.fromJson(json))
            .toList();
      }
      throw Exception('Invalid response format');
    } catch (e) {
      throw Exception('Error fetching rental vehicles: $e');
    }
  }

  Future<RentalVehicle> getRentalVehicleById(int id, {String? token}) async {
    try {
        final response = await _apiService.get('/rental-vehicles/$id', token: token);
        return RentalVehicle.fromJson(response as Map<String, dynamic>);
      } catch (e) {
        throw Exception('Error fetching rental vehicle: $e');
    }
  }

  Future<RentalVehicle> createRentalVehicle(
      Map<String, dynamic> data, String token) async {
    try {
        final response = await _apiService.post('/rental-vehicles', data, token: token);
        return RentalVehicle.fromJson(response as Map<String, dynamic>);
      } catch (e) {
        throw Exception('Error creating rental vehicle: $e');
    }
  }

  Future<RentalVehicle> updateRentalVehicle(
      int id, Map<String, dynamic> data, String token) async {
    try {
        final response = await _apiService.patch('/rental-vehicles/$id', data, token: token);
        return RentalVehicle.fromJson(response as Map<String, dynamic>);
      } catch (e) {
        throw Exception('Error updating rental vehicle: $e');
    }
  }

  Future<void> deleteRentalVehicle(int id, String token) async {
    try {
      await _apiService.delete('/rental-vehicles/$id', token: token);
    } catch (e) {
      throw Exception('Error deleting rental vehicle: $e');
    }
  }
}

File: frontend/src/app/services/rental-vehicles.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';
import { environment } from '../../environments/environment';

export interface RentalVehicle {
  id: number;
  company_name: string;
  vehicle_type: 'automovil' | 'camioneta';
  vehicle_name: string;
  characteristics: string;
  photo_url?: string;
  whatsapp_number: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface RentalVehicleCreate {
  company_name: string;
  vehicle_type: 'automovil' | 'camioneta';
  vehicle_name: string;
  characteristics: string;
  photo_url?: string;
  whatsapp_number: string;
}

export interface RentalVehicleUpdate {
  company_name?: string;
  vehicle_type?: 'automovil' | 'camioneta';
  vehicle_name?: string;
  characteristics?: string;
  photo_url?: string;
  whatsapp_number?: string;
  is_active?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class RentalVehiclesService {
  private apiUrl = `${environment.apiUrl}/rental-vehicles`;

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) { }

  private getHeaders(): HttpHeaders {
    const token = this.authService.getToken();
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
  }

  // Get all rental vehicles
  getAllRentalVehicles(isActive?: boolean): Observable<RentalVehicle[]> {
    let url = this.apiUrl;
    if (isActive !== undefined) {
      url += `?is_active=${isActive}`;
    }
    return this.http.get<RentalVehicle[]>(url, { headers: this.getHeaders() });
  }

  // Get single rental vehicle by id
  getRentalVehicleById(id: number): Observable<RentalVehicle> {
    return this.http.get<RentalVehicle>(`${this.apiUrl}/${id}`, { headers: this.getHeaders() });
  }

  // Create rental vehicle
  createRentalVehicle(data: RentalVehicleCreate): Observable<RentalVehicle> {
    return this.http.post<RentalVehicle>(this.apiUrl, data, { headers: this.getHeaders() });
  }

  // Update rental vehicle
  updateRentalVehicle(id: number, data: RentalVehicleUpdate): Observable<RentalVehicle> {
    return this.http.patch<RentalVehicle>(`${this.apiUrl}/${id}`, data, { headers: this.getHeaders() });
  }

  // Delete rental vehicle
  deleteRentalVehicle(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`, { headers: this.getHeaders() });
  }
}
